specs/agents/sanitization.md

```diff
@@ -1,16 +1,58 @@
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to
+be interpreted as described in
+[RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
 ## Data sanitization
 
 ### `sanitize_field_names` configuration
 
 Sometimes it is necessary to sanitize, i.e., remove,
 sensitive data sent to Elastic APM.
-This config accepts a list of wildcard patterns of field names which should be sanitized.
-These apply to HTTP headers (including cookies) and `application/x-www-form-urlencoded` data (POST form fields).
-The query string and the captured request body (such as `application/json` data) will not get sanitized.
+
+This config accepts a list of wildcard patterns of field names which control
+how an agent will sanitize data.
 
 |                |   |
 |----------------|---|
 | Type           | `List<`[`WildcardMatcher`](../../tests/agents/json-specs/wildcard_matcher_tests.json)`>` |
 | Default        | `password, passwd, pwd, secret, *key, *token*, *session*, *credit*, *card*, authorization, set-cookie` |
 | Dynamic        | `true` |
 | Central config | `true` |
+
+## Configuration
+
+Agents MUST provide a minimum default configuration of
+
+    [ 'password', 'passwd', 'pwd', 'secret', '*key', '*token*', '*session*',
+      '*credit*','*card*', 'authorization', 'set-cookie']
+
+for the `sanitize_field_names` configuration value.  Agent's MAY include the
+following extra fields in their default configuration to avoid breaking changes
+
+    ['pw','pass','connect.sid']
+
+If an end-user configures different values, these values MUST **replace** the
+above values. An end-user's configured values MUST NOT be merged with these
+default values.
+
+## Sanitizing Values
+
+If a payload field's name (a header key, a form key) matches a configured
+wildcard, that field's _value_ MUST be removed/redacted and the key itself
+MUST still be reported in the agent payload. Agents MAY chose the string
+they use to replace the value so long as it's consistent and does not reveal
+the value it has replaced. Some example replacement strings
+include `REDACTED`, `**********`, `-----------`, etc.
+
+Fields that MUST be sanitized are the HTTP Request headers, HTTP Response
+headers, and form fields in an `application/x-www-form-urlencoded` request
+body.  No fields (including `set-cookie` headers) are exempt from this.
```

Comment:
There's also [`transaction.context.request.cookies`](https://github.com/elastic/apm-server/blob/6062235783c398c61722ffb23c98ab790bcdf6e7/docs/spec/v2/transaction.json#L129-L135) which is a map of cookie names to values. This map must be redacted, too.

Reply:
Only transaction.context.request.cookies map should be filtered or those `Set-Cookie` HTTP header(s) that reference excluded cookie names as well (assuming that the configured value is different from the default and `Set-Cookie` is not included)?

Reply:
`Set-Cookie` is a response header. The corresponding request header is `Cookie`. As the content of the `Cookie` header is the same as `transaction.context.request.cookies`, in the Java agent, we're removing `Cookie` from `transaction.context.request.headers`. This both minimizes the amount of duplicate data and eliminates the complication of having to redact parts of a header value. 

For response headers, there's no equivalent to `transaction.context.request.cookies`, so we just sanitize the `transaction.context.response.headers` map.

See also https://github.com/elastic/apm-agent-java/blob/master/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/SanitizingWebProcessor.java

Reply:
> in the Java agent, we're removing `Cookie` from `transaction.context.request.headers`. 

Should we include it in the spec?

Reply:
@felixelastic @SergeyKleyman @basepi If I'm reading the above correctly, it sounds like there's three data structures/values to consider here

1. The `Set-Cookie` response header
2. The `Cookie` request header
3. A representation of the key/values pairs of a `Set-Cookie` header that's stored in `transaction.context.request.cookies`

If I spec-ed this as something like

> MUST redact `transaction.context.request.cookies`

> If agents capture values for `Set-Cookie` or `Cookie` request/response headers, it MUST redact the field values here as well.

would that capture what we're after? Or is there more nuance here?

Reply:
The relevant data structures are:
* `transaction.context.request.cookies`
* `transaction.context.request.headers`
* `transaction.context.request.body` (if form/url encoded)
* `transaction.context.response.headers`

Also the `Cookie` header should be removed form `transaction.context.request.headers` as it's already recorded in `transaction.context.request.cookies` and because it would sanitization quite difficult/expensive.

Reply:
> Also the `Cookie` header should be removed form `transaction.context.request.headers`

Instead of hardcoding `Cookie` header removal can this behavior be achieved by adding `cookie` to `SANITIZE_FIELD_NAMES` default value?

Reply:
@SergeyKleyman Re:

> can this behavior be achieved by adding cookie to SANITIZE_FIELD_NAMES default value?

Good question! Thinking that through --  I'm not sure this would achieve what we want. Here's my reasoning: If we added `cookie` to `SANITIZE_FIELD_NAMES`, we'd be instructing the agent to _REDACT_ the a header named cookie, and not to remove it completely.  

Also, I've been operating under the assumption that the default fields were decided on a while ago and would be a pain to change now after some agent teams have already landed this functionality or are midstream on it.

Reply:
@felixbarny Thank you for sharing those data structures -- I learned something. It looks like the Node.js Agent doesn't capture `transaction.context.request.cookies` at all.  Wondering if we need a spec for what part of the request gets captured or if that's just going spec wild. 

More pertinent to our discussion though -- What do you think about this as final language for future proofing things?

> Fields that MUST be sanitized are the HTTP Request headers, HTTP Response
headers, and form fields in an `application/x-www-form-urlencoded` request
body.  Additionally, if cookie headers are parsed into name/value pairs and reported 
to APM Server via the agent (for example, `transaction.context.request.cookies`), the 
cookie names must be sanitized as well.

Reply:
Ah, I didn't know some agents don't send `request.cookies`. 
That property is spec'd in the [intake API](https://github.com/elastic/apm-server/blob/d18c559e4136b2213a4f6e206258c5d303ec1b0f/docs/spec/v2/transaction.json#L129-L135) but it's not a required field.

I like that wording!

Reply:
> Here's my reasoning: If we added `cookie` to `SANITIZE_FIELD_NAMES`, we'd be instructing the agent to _REDACT_ the a header named cookie, and not to remove it completely.

The main reason for my suggestion is that it might look strange to the users that `Set-Cookie` header is redacted while `Cookie` header is removed completely.

Reply:
> Additionally, if cookie headers are parsed into name/value pairs and reported 
to APM Server via the agent (for example, `transaction.context.request.cookies`), the 
cookie names must be sanitized as well.

It might be worth adding something like:

> In addition parsed cookie headers MUST NOT be included in the set of headers (in the example above in `transaction.context.request.headers`)